lib/srv/reexec.go

```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+			var req networking.Request
+			if err := json.Unmarshal(buf[:n], &req); err != nil {
+				slog.With("error", err).ErrorContext(ctx, "Error parsing networking request.")
+				return
+			}
+			log := slog.With("request", req)
+
+			if fn == 0 {
+				log.ErrorContext(ctx, "Networking request requires a control file.")
+				return
+			}
+
+			controlConn, err := uds.FromFile(fbuf[0])
+			_ = fbuf[0].Close()
+			if err != nil {
+				log.ErrorContext(ctx, "Failed to get a connection from control file.")
+				return
+			}
+			defer controlConn.Close()
+
+			log.Debug("Handling networking request")
+			netFile, err := handleNetworkingRequest(ctx, req)
+			if err != nil {
+				log.With("error", err).ErrorContext(ctx, "Error handling networking request.")
+				if _, err2 := controlConn.Write([]byte(err.Error())); err2 != nil {
+					log.With("error", err2.Error()).ErrorContext(ctx, "Failed to write error to control conn.")
+				}
+			}
+			defer netFile.Close()
+
+			if _, _, err := uds.WriteWithFDs(controlConn, nil, []*os.File{netFile}); err != nil {
+				log.With("error", err.Error()).ErrorContext(ctx, "Failed to write networking file to control conn.")
 			}
 		}()
 	}
 }
 
-// RunLocalForward reads in the command to run from the parent process (over a
-// pipe) then port forwards.
-func RunLocalForward() (errw io.Writer, code int, err error) {
-	errw, code, err = runForward(handleLocalPortForward)
-	return errw, code, trace.Wrap(err)
+func handleNetworkingRequest(ctx context.Context, req networking.Request) (*os.File, error) {
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	switch req.Operation {
+	case networking.NetworkingOperationDial:
+		var d net.Dialer
+		conn, err := d.DialContext(ctx, req.Network, req.Address)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer conn.Close()
+
+		connFD, err := getConnFile(conn)
+		return connFD, trace.Wrap(err)
+
+	case networking.NetworkingOperationListen:
+		listener, err := newListener(ctx, req.Network, req.Address)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer listener.Close()
+
+		listenerFD, err := getListenerFile(listener)
+		return listenerFD, trace.Wrap(err)
+
+	case networking.NetworkingOperationListenAgent:
+		// Create a temp directory to hold the agent socket.
+		sockDir, err := os.MkdirTemp(os.TempDir(), "teleport-")
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		socketPath := filepath.Join(sockDir, fmt.Sprintf("teleport-%v.socket", os.Getpid()))
+
+		listener, err := newListener(ctx, "unix", socketPath)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer listener.Close()
+
+		listenerFD, err := getListenerFile(listener)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Update the agent forwarding socket with more restrictive permissions.
+		if err = os.Chmod(socketPath, teleport.FileMaskOwnerOnly); err != nil {
```

Comment:
If this is necessary, this is vulnerable to a connect() between listen() and chmod(); the directory containing the socket should be created with restrictive permissions instead.

Reply:
You're right, this is how ssh does it too.